torchtitan/optimizer.py

```diff
@@ -121,44 +156,30 @@ def linear_warmup_linear_decay(
 class SchedulersContainer:
     """Util for calling step on multiple learning rate schedulers needed for virtual pipeline stages"""
 
-    def __init__(self, optimizers, lr_lambda):
+    def __init__(self, optimizers, lr_lambda) -> None:
         self.schedulers = []
         for optimizer in optimizers:
             self.schedulers.append(LambdaLR(optimizer, lr_lambda=lr_lambda))
 
-    def step(self):
+    def step(self) -> None:
         for schedulers in self.schedulers:
             schedulers.step()
 
+    def get_lr_scheduler_state(self) -> Dict[str, Any]:
+        state_dict = {}
+        if len(self.schedulers) == 1:
+            state_dict["lr_scheduler"] = self.schedulers[0]
+        else:
+            # For now, pipeline-parallel with looped schedules does not support resharding for lr_scheduler.
+            # It should only support saving and loading a distributed checkpoint with the same number of pp ranks
```

Comment:
I feel this could be solved, if we do flattening on lr_scheduler state_dict similar to what we did to models and optimizers.

Reply:
This is hard to achieve. LRScheduler's state_dict is very unstructured. It basically just returns `self.__dict__.items()`, which can be anything. And `LRScheduler` doesn't define a parameter group structure. So different `LRScheduler` may have different implementations. I'm not sure how to flatten `LRScheduler` in a general approach. Unless we only focus on one `LRScheduler`.

Reply:
@fegin I see. Then maybe we can just focus on the LambdaLR, which should be straightforward -- since every scheduler has the same schedule, we can store only one state and recreate schedulers for each optimizer on the fly when doing checkpoint loading.

Reply:
ye, I want to emphasize that the direction is correct to flatten LRScheduler, which TorchRec also does, iirc. But because we may not have bandwidths to support all of them, so focusing on one or two and claim what TorchTitan supports is a good idea.